src/DataTypes/Serializations/SerializationTuple.cpp

```diff
@@ -119,6 +119,11 @@ static ReturnType addElementSafe(size_t num_elems, IColumn & column, F && impl)
     return ReturnType(true);
 }
 
+void SerializationTuple::readElementsSafe(DB::IColumn & column, std::function<void()> && read_func)
+{
+    addElementSafe<void>(assert_cast<ColumnTuple &>(column).getColumns().size(), column, [&](){ read_func(); return true; });
```

Comment:
Isn't it the same logic in `IRowInputFormat` ? https://github.com/ClickHouse/ClickHouse/blob/a0a96b5580d21cb2791526f727a5bc48468c87ee/src/Processors/Formats/IRowInputFormat.cpp#L194-L201

Reply:
Yes, it's similar, but logic in `IRowInputFormat` is not enough, we call `popBack` on top level column, for example, on `ColumnTuple`, but if we didn't serialize it safely, nested tuple column may have different number of rows, and `ColumnTuple::popBack` won't restore initial state correctly